### ISSUE TITLE
Adding pdiff to tests suite

### ISSUE THREAD

**Issue by pierrejoye** [libgd/libgd#216]
Adding a port of pdiff (bundled) to the tests suite to better fit our needs in the testing area. pdiff is as of today the best available tool for this purpose and can be used when work with compression or other operations where approximation may result in slight differences while visually the same.

It is GPL but it is fine as long as it is only part of the testing suite.

**Comment by cmb69** [comment]
I'm not sure if [pdiff](http://pdiff.sourceforge.net/) would actually suit *our* needs. I'm afraid it would report too many false positives due to its advanced algorithm. Also [pHash](http://www.phash.org/) would probably be unsuitable.

Actually, we're mostly interested in *pixel-exact* operations. The only exceptions would be testing of the lossy output formats, and some precision/rounding issues (particularly, the freetype rending appears to be affected). The latter should ideally be solved, though.

Anyhow, I wonder if we won't come a long way with a simple [euclidean distance calculation](http://www.math.ttu.edu/outreach/sumac/pics/lhs.pdf). Besides that it's easy to implement (and maintain), it would primarily check for pixel exactness, but also allow for some differences (according to some given threshold).

**Comment by cmb69** [comment]
I had a closer look at pdiff and an alternative Euclidean distance comparison, see <https://cmb69.github.io/pdiff/index.html>.

TL;DR: I don't think that we should add pdiff to our test framework.

**Comment by pierrejoye** [comment]
Why not?

Euclidean may not fit for operations like filter or geometrical
transformations.

Also Cairo, which is all about rastering, do use pdiff. I fail to
understand why it would not math for us. Do you have cases why it does not
fit?

**Comment by cmb69** [comment]
> Do you have cases why it does not fit?

See the examples and my conclusion on <https://cmb69.github.io/pdiff/index.html>.

**Comment by pierrejoye** [comment]
@cmb69 is right. We can revisit this topic if needed when usable alternatives exist.
